lib/src/widgets/yaru_search_field.dart

```diff
@@ -0,0 +1,351 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+import 'package:yaru/yaru.dart';
+import 'package:yaru_icons/yaru_icons.dart';
+import 'package:yaru_widgets/constants.dart';
+import 'package:yaru_widgets/src/widgets/yaru_icon_button.dart';
+
+/// A [TextField] to with fully rounded corners,
+/// ideally in a [YaruWindowTitleBar] or [YaruDialogTitleBar]
+class YaruSearchField extends StatefulWidget {
+  const YaruSearchField({
+    super.key,
+    this.text,
+    this.onSubmitted,
+    this.hintText,
+    this.height = kYaruTitleBarItemHeight,
+    this.contentPadding = const EdgeInsets.only(
+      bottom: 10,
+      top: 10,
+      right: 15,
+      left: 15,
+    ),
+    this.autofocus = true,
+    this.onClear,
+    this.onChanged,
+    this.radius = const Radius.circular(kYaruTitleBarItemHeight),
+    this.style = YaruSearchFieldStyle.filled,
+    this.borderColor,
+    this.fillColor,
+  });
+
+  /// Optional [String] forwarded to the internal [TextEditingController]
+  final String? text;
+
+  /// Optional [String] used inside the internal [InputDecoration]
+  final String? hintText;
+
+  /// The callback forwarded to the [TextField] used when the enter key is pressed
+  final void Function(String? value)? onSubmitted;
+
+  /// The callback forwarded to the [TextField] used when input changes
+  final void Function(String value)? onChanged;
+
+  /// Optional callback used to clear the [TextField]. If provided an [IconButton] will use it
+  /// as the suffix icon inside the [InputDecoration]
+  final void Function()? onClear;
+
+  /// The height of the [TextField] that defaults to [kYaruTitleBarItemHeight]
+  final double height;
+
+  /// The padding for the [InputDecoration] that defaults to `EdgeInsets.only(bottom: 10,top: 10, right: 15, left: 15)`
+  final EdgeInsets contentPadding;
+
+  /// Defines if the [TextField] is autofocused on build
+  final bool autofocus;
+
+  /// Defines the radius for the corners.
+  final Radius radius;
+
+  final YaruSearchFieldStyle style;
+
+  /// Optional [Color] for the border. If not provided
+  /// it will fall back to `ColorScheme.dividerColor`.
+  final Color? borderColor;
+
+  /// Optional [Color] for the border. If not provided
+  /// it will fall back to `ColorScheme.dividerColor`.
+  final Color? fillColor;
+
+  @override
+  State<YaruSearchField> createState() => _YaruSearchFieldState();
+}
+
+class _YaruSearchFieldState extends State<YaruSearchField> {
+  late TextEditingController _controller;
+  final _focusNode = FocusNode();
```

Comment:
i have a strong feeling we should allow users to pass both or else they won't be able to control selection or focus

Reply:
yes you are right, there is something f*cked up with the focus on rebuild

Reply:
you think that those fields should be completely out here and make the field totally stateless?

Reply:
hiding the clear button now if _controller.text is empty

Reply:
they are usually only needed for more advanced use cases. it might be inconvenient to always require them. i'd follow the common practice of creating and disposing them here if the user didn't supply them.

Reply:
ok!

Reply:
like this?

Reply:
what could possibly go wrong when suggesting changes written on the fly via github's web GUI :)

Reply:
haha :laughing: well fixed it locally now :smile: